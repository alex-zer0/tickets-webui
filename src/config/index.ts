const config = {
  apiUrl: 'http://localhost:8000/api',
  pusherTicketsChannel: 'tickets_channel',
  pusherTicketsEvent: 'tickets_events',
  pusherAppKey: '9a46d976311a4b4ae129',
  pusherCluster: 'eu',
};

export default config;
