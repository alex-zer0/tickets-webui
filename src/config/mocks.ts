export const fakeEvent = () => ({
  name: 'Event',
  url: '/event',
  bannerUrl: 'http://event-banner-url',
  date: '2020-10-10 00:00:00',
  location: 'Russia, Saint-Petersburg',
  description: 'Test Description',
  tickets: [],
});

export const fakeTickets = () => [
  {id: 1, name: 'VIP', available: 10},
  {id: 2, name: 'Standard', available: null},
  {id: 3, name: 'Alumnus', available: null},
];
