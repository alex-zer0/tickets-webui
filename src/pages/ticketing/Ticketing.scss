.Ticketing {
    box-shadow: 0 1px 4px rgba(0,0,0,.3);
    width: 900px;
    margin: 0 auto;

    & > section {
        padding: 1rem 1.5rem;
    }
}
