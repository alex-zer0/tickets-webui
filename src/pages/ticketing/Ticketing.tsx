import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router';
import { Dispatch } from 'redux';
import Header from 'src/components/header/Header';
import { checkoutEventTickets, fetchEventByUrl } from 'src/store/actions/events';
import { EventInfo, EventsState, EventTicket, TicketFormEventTicket } from 'src/store/reducers/events';
import Loading from 'src/components/Loading';
import { formatDate } from 'src/services/date.service';
import './Ticketing.scss';
import TicketsForm from './tickets-form/TicketsForm';

export interface TicketingProps {
  event: EventInfo | null
  loading: boolean
  checkoutTickets: TicketFormEventTicket[] | null
  match: {
    params: {[key: string]: string},
  }
  checkoutEventTickets: (url: string, tickets: TicketFormEventTicket[]) => void
  fetchEvent: (url: string) => void
}

export class Ticketing extends Component<TicketingProps> {
  public componentDidMount() {
    const { match, fetchEvent } = this.props;
    if (match) {
      const { params: { eventUrl } } = match;
      fetchEvent(eventUrl);
    }
  }

  public handleCheckout = (tickets: EventTicket[]) => {
    const { checkoutEventTickets, event } = this.props;
    if (!event) {
      return;
    }
    checkoutEventTickets(event.url, tickets);
  }

  public renderTicketsForm(tickets?: EventTicket[]) {
    if (!tickets || tickets.length === 0) {
      return null;
    }
    return (
      <TicketsForm
        tickets={tickets}
        checkout={this.handleCheckout}
      />
    );
  }

  public render() {
    const { loading, event, checkoutTickets } = this.props;
    if (checkoutTickets) {
      return <Redirect to='/checkout' />;
    }
    if (loading) {
      return <Loading />;
    }
    if (!event) {
      return <div>There is no Event with such URL</div>;
    }
    const { name, date, location, description, bannerUrl, tickets } = event;
    return (
      <div className='Ticketing'>
        <Header title={name} imageUrl={bannerUrl} />
        <section>
            <p>{formatDate(date)}</p>
            <p>{location}</p>
            {this.renderTicketsForm(tickets)}
            <footer>
              <h5>Description</h5>
              <p>{description}</p>
            </footer>
        </section>
      </div>
    );
  }
}

function mapStateToProps(state: { eventsState: EventsState }) {
  return {
    loading: state.eventsState.eventLoading,
    event: state.eventsState.activeEvent,
    checkoutTickets: state.eventsState.checkoutTickets
  };
}

function mapDispatchToProps(dispatch: Dispatch) {
  return {
    fetchEvent: (url: string) => dispatch(fetchEventByUrl(url)),
    checkoutEventTickets: (url: string, tickets: TicketFormEventTicket[]) => dispatch(checkoutEventTickets(url, tickets)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Ticketing);
