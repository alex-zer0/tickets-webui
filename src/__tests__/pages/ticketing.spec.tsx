import React from 'react';
import renderer from 'react-test-renderer';
import { fakeEvent } from '../../config/mocks';
import { Ticketing, TicketingProps } from '../../pages/ticketing/Ticketing';
import { BrowserRouter } from 'react-router-dom';

const ticketingProps: TicketingProps = {
  event: fakeEvent(),
  checkoutTickets: null,
  loading: false,
  match: { params: {} },
  checkoutEventTickets: jest.fn(),
  fetchEvent: jest.fn()
};

describe('PAGES: Ticketing', () => {
  it('renders Loader', () => {
    const props = {
      ...ticketingProps,
      loading: true,
    };
    const tree = renderer
      .create(<BrowserRouter><Ticketing {...props}/></BrowserRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders w/o exceptions', () => {
    const tree = renderer
      .create(<BrowserRouter><Ticketing {...ticketingProps}/></BrowserRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
