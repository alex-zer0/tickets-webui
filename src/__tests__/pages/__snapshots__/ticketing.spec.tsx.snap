// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PAGES: Ticketing renders Loader 1`] = `
<div>
  Loading...
</div>
`;

exports[`PAGES: Ticketing renders w/o exceptions 1`] = `
<div
  className="Ticketing"
>
  <header
    className="Header"
  >
    <img
      alt="nodejs"
      className="Header__image"
      src="http://event-banner-url"
    />
    <h1
      className="Header__title"
    >
      Event
    </h1>
  </header>
  <section>
    <p>
      Invalid DateTime
    </p>
    <p>
      Russia, Saint-Petersburg
    </p>
    <footer>
      <h5>
        Description
      </h5>
      <p>
        Test Description
      </p>
    </footer>
  </section>
</div>
`;
